Generated hint for @ContributesBinding and @ContributesTo explicitly defined as internal to support Kotlin 1.4 Explicit API Mode

marcellogalhardo opened the pull request:
When using Anvil in a module that has strict API Mode enable, `ContributesBindingGenerator` and `ContributesToGenerator` will generate the hint used internally for merging modules and components interface without a visibility modifier and therefore, Kotlin Explicit API will fail the compilation.

Fixes #140.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/square/anvil?pullRequest=141) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/square/anvil?pullRequest=141) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/square/anvil?pullRequest=141) it.</sub>

marcellogalhardo commented:
I'm not sure what went wrong, I will investigate it.

marcellogalhardo commented:
I'm investigating the failing tests. 🤔

marcellogalhardo commented:
I realised my previous solution was unnecessary as these hints are usually used internally. Therefore, `internal` looks like the right visibility modifier to be used here. I also rebased my branch on top of last main branch. All tests are now green.

I updated the PR description.

vRallev reviewed (changes requested):
> I realised these generated hints are used internally only. Therefore, `internal` looks like the right visibility modifier to be used here.

That's not correct. `internal` means this code is only used inside of this module and is not visible outside of this modules. However, Anvil needs to see the generated code in other modules. The generated hints must be public.

marcellogalhardo commented:
My understand was that these values are loaded directly from the GeneratedHint file and the modifiers didn't affect Anvil but I have missed something.

I changed the code to use `public` visibility modifier but Explicit API Mode demands to set a return type. This made [ModuleMergerTest.'inner modules in a merged component fail'](https://github.com/marcellogalhardo/anvil/blob/main/compiler/src/test/java/com/squareup/anvil/compiler/ModuleMergerTest.kt#L713) fail and while investigating I concluded that:

_Before the changes:_ when using a inner class without a return type `ContributedHint.reference` was a `DeferredType` that would be evaluate only when trying to access it ([ClassScanner.findContributedClasses](https://github.com/marcellogalhardo/anvil/blob/main/compiler/src/main/java/com/squareup/anvil/compiler/ClassScanner.kt#L55)) causing an error and an [ExitCode.INTERNAL_ERROR](https://github.com/marcellogalhardo/anvil/blob/main/compiler/src/test/java/com/squareup/anvil/compiler/ModuleMergerTest.kt#L729). From the stack trace:
```
	...
	at org.jetbrains.kotlin.types.DeferredType.getDelegate(DeferredType.java:107)
	at org.jetbrains.kotlin.types.WrappedType.getArguments(KotlinType.kt:129)
	at com.squareup.anvil.compiler.UtilsKt.argumentType(Utils.kt:116)
	at com.squareup.anvil.compiler.ContributedHint.toClassDescriptor(ClassScanner.kt:87)
	at com.squareup.anvil.compiler.ContributedHint.access$toClassDescriptor(ClassScanner.kt:68)
	at com.squareup.anvil.compiler.ContributedHint$reference$2.invoke(ClassScanner.kt:72)
	...
```

_After the changes:_ with a return type declared, the `ContributedHint.reference` will always have a type defined (e.g., `KClass<InnerModule>`) and no `DeferredType` will be used. The inner module validation would still happen as expected causing a `ExitCode.COMPILATION_ERROR`. I didn't find any other difference.

The other details of the test are still the same. And the right exception is still throw:

```
test/Source.kt: (10, 18): com.squareup.anvil.compiler.AnvilCompilationException: Back-end (JVM) Internal error: It seems like you tried to contribute an inner class to its outer class. This is not supported and results in a compiler error.
File being compiled: (10,18) in /var/folders/b4/131t8pg542j3q_1g7c0b62wc0000gp/T/Kotlin-Compilation7934599964874323027/sources/src/main/java/com/squareup/test/Source.kt
The root cause java.lang.IllegalStateException was thrown at: org.jetbrains.kotlin.types.ErrorUtils$1.getPackage(ErrorUtils.java:95)
```

I changed the test to expect an `ExitCode.COMPILATION_ERROR` but I'm not sure if this feels wrong and I don't see a way to keep the same `DeferredType` behaviour while with a return type declared...

What do you think @vRallev? 🤔 

Sorry for the long message.

vRallev commented:
Yes, that is strange. But I don't see why this behavior changed. The original message is still preserved, so I'm not concerned.

vRallev commented:
Note that I created #144 as a follow-up.
